codegen/src/utils/fetch_metadata.rs

```diff
@@ -33,25 +36,75 @@ fn tokio_block_on<T, Fut: std::future::Future<Output = T>>(fut: Fut) -> T {
 
 /// Returns the metadata bytes from the provided URL.
 pub async fn fetch_metadata_bytes(url: &Uri) -> Result<Vec<u8>, FetchMetadataError> {
-    let hex = fetch_metadata_hex(url).await?;
-    let bytes = hex::decode(hex.trim_start_matches("0x"))?;
-    Ok(bytes)
-}
-
-/// Returns the raw, 0x prefixed metadata hex from the provided URL.
-pub async fn fetch_metadata_hex(url: &Uri) -> Result<String, FetchMetadataError> {
-    let hex_data = match url.scheme_str() {
+    let bytes = match url.scheme_str() {
         Some("http") | Some("https") => fetch_metadata_http(url).await,
         Some("ws") | Some("wss") => fetch_metadata_ws(url).await,
         invalid_scheme => {
             let scheme = invalid_scheme.unwrap_or("no scheme");
             Err(FetchMetadataError::InvalidScheme(scheme.to_owned()))
         }
     }?;
+
+    Ok(bytes)
+}
+
+/// Returns the raw, 0x prefixed metadata hex from the provided URL.
+pub async fn fetch_metadata_hex(url: &Uri) -> Result<String, FetchMetadataError> {
+    let bytes = fetch_metadata_bytes(url).await?;
+    let hex_data = format!("0x{}", hex::encode(bytes));
     Ok(hex_data)
 }
 
-async fn fetch_metadata_ws(url: &Uri) -> Result<String, FetchMetadataError> {
+/// Execute runtime API call and return the specified runtime metadata version.
+async fn fetch_latest_metadata(client: impl ClientT) -> Result<Vec<u8>, FetchMetadataError> {
+    const V15_METADATA_VERSION: u32 = u32::MAX;
+    let bytes = V15_METADATA_VERSION.encode();
+
+    // Runtime API arguments are scale encoded hex encoded.
+    let version: String = format!("0x{}", hex::encode(&bytes));
+
+    // Returns a hex(Option<frame_metadata::OpaqueMetadata>).
+
+    let result: Result<String, _> = client
+        .request(
+            "state_call",
+            rpc_params!["Metadata_metadata_at_version", &version],
+        )
+        .await;
+
+    if let Ok(raw) = result {
+        let raw_bytes = hex::decode(raw.trim_start_matches("0x"))?;
+
+        let opaque: Option<frame_metadata::OpaqueMetadata> = Decode::decode(&mut &raw_bytes[..])?;
+        let bytes = opaque.ok_or(FetchMetadataError::Other(
+            "Metadata version not found".into(),
+        ))?;
+
+        return Ok(bytes.0);
+    }
+
+    // The `Metadata_metadata_at_version` RPC failed, fall back to the old `state_getMetadata`.
+    let raw: String = client.request("state_getMetadata", rpc_params![]).await?;
+
+    let raw_bytes = hex::decode(raw.trim_start_matches("0x"))?;
+    // Decode the metadata to inspect the RuntimeMetadata variant.
+    let meta: RuntimeMetadataPrefixed = Decode::decode(&mut &raw_bytes[..])?;
+    // Update available versions to V15.
+    let metadata = match meta.1 {
+        RuntimeMetadata::V14(v14) => metadata_v14_to_latest(v14),
```

Comment:
Doesn't this mean that we'll always be returning V15 metadata from a node in a bunch of our CLI commands?

As you know, I'm keen to make sure that whereveer possible, the CLI commands (eg `subxt metadata`) returns whatever the node has available rather than modifying anything (unless we need to strip pallets or whatever, which will cause an upgrade).

I guess I'd be keen for this `metadata_v14_to_latest` call to be explicit in whatever command requires the upgrade, rather than buried away deeper in the fetch logic, so that it's super clear where exactly we are upgrading it :)

Reply:
Yep, that makes sense! Will modify this to return:
- V15 as `RuntimeMetadataPrefixed` if the `Metadata_metadata_at_version(u32::MAX)` exists
- V14 (or latest stable) as `RuntimeMetadataPrefixed` via `Metadata_metadata` runtime call 

We can then upgrade it later in the codegen when we generate the interface